### PR TITLE
MAINT: Switch to sphinx-theme-builder

### PR DESCRIPTION
**Pull request by AakashGfude**
This PR switches to using [sphinx-theme-builder](https://github.com/pradyunsg/sphinx-theme-builder), following this issue https://github.com/executablebooks/sphinx-book-theme/issues/417

- [x] Use sphinx-theme-builder as the build backend
- [x] Use webpack to compile and build assets. Run stb compile.
- [x] Build docs.
- [x] Run tests.
- [x] Do a thorough cleanout
- [x] Update Docs

### CODE REVIEW

File: noxfile.py
Comment (choldgraf) [inline noxfile.py:1]: What was your reasoning for moving to `nox`? I personally prefer `nox` to `tox` as I think it is more useful for actions outside of testing, but wondering why you made the switch as a part of this PR.

File: package.json
Comment (choldgraf) [inline package.json:19]: Some of these we can likely clean out (e.g. we do not test with `pa11y-ci`). Let's shoot for the minimal possible dependencies

**Review by choldgraf — COMMENTED** [review]
A couple quick thoughts - thanks!

File: package.json
Comment (AakashGfude) [reply package.json:19]: yes. need to do a thorough clean out.

File: noxfile.py
Comment (AakashGfude) [reply noxfile.py:1]: Similar reasoning. It is more pythony, and we can run python code inside it to aid testing. Since pydata-sphinx-theme was also using it, so jumped at the opportunity 😬 . Making theme package structures more consistent. But happy to continue with tox, if we conclude its better for this project.

Comment (choldgraf) [reply noxfile.py:1]: Well I am personally a fan of `nox` for the following reasons:

- I find `noxfile.py` is easier to quickly read and understand what's going on, since it looks "like a Python script". This also likely is more accessible to more people that aren't familiar with software development workflows.
- I find it easier to re-use commands (like the install command) since you just write them as python functions

We don't have an official org policy on using tox vs. nox (at least I couldn't find anything in [the development conventions](https://executablebooks.org/en/latest/contributing.html)), so I am fine to try it out here and get feedback from the team over time on whether they like it or not. If worst comes to worst we can always switch back to `tox`.

Comment (chrisjsewell) [reply noxfile.py:1]: It's not the end of the world, but I am against using nox.
tox is by far the more commonly used package, and declarative configuration is much preferred over python files. Plus I use tox-conda which there is no equivalent plugin system for nox.
tox is the org policy and is implemented in nearly every repo in EBP. nox shouldn't only be used if there is a clear requirement for more flexible configuration.

Comment (chrisjsewell) [reply noxfile.py:1]: I'd also note that this is undoing #211, where we already had this discussion, and I'd moved the package from nox to tox

Comment (choldgraf) [reply noxfile.py:1]: Maybe this is overkill, but we could support both for a time? As long as we document each it shouldn't be a huge deal, and maybe an opportunity to compare their use a bit. Those files don't change much 🤷.

This feels like one of those things that there is no "right" answer for 🤔. I don't have any problems with tox so if push comes to shove I'm fine using that, but for some reason nox just feels more user friendly to me, I can't put my finger on it.

Comment (chrisjsewell) [reply noxfile.py:1]: So are you advocating for `conf.py` over `_config.yml` also 😝 
Again, I don't want to bike-shed, it should be what is most helpful for people actively doing the development on the package. But, I would most certainly block the change, if what on any of the repos I am actually actively working on 😬

Comment (AakashGfude) [reply noxfile.py:1]: @choldgraf @chrisjsewell Since both seem to do our work just fine. I will just keep tox in this PR, and we can try nox in the later PRs if we want to.

Comment (pradyunsg) [reply noxfile.py:1]: I'd suggest keeping this change separate. There's a decent amount of baggage in the nox vs tox topic in this project, and it'd be best to keep it separated.

I'm also on board with nox (see pip's development processes) but that's besides the point.

Comment (choldgraf) [reply noxfile.py]: I agree with @pradyunsg - no need to conflate two different changes in one PR that is already touching a lot of stuff.

**Comment by AakashGfude** [comment]
@pradyunsg where do you suggest something like a https://github.com/executablebooks/sphinx-book-theme/tree/master/sphinx_book_theme/translations folder should go in terms of the layout system according to `sphinx-theme-builder` https://sphinx-theme-builder.readthedocs.io/en/latest/reference/filesystem-layout/ .

**Comment by choldgraf** [comment]
I think we could just put it under theme/sphinx_book_theme/translations

**Comment by AakashGfude** [comment]
> I think we could just put it under `theme/sphinx_book_theme/translations`

@choldgraf since technically we are compiling jsons to *.mo files and and `sphinx-theme-builder` technically encourages any assets that need to be compiled to be put in `src/sphinx_book_theme/assets`, so I am a bit confused. Also, I think we should add the creation of *.mo files in webpack instead of committing them?

**Comment by pradyunsg** [comment]
> we are compiling jsons to *.mo files, and sphinx-theme-builder technically encourages any assets that need to be compiled to be put in src/sphinx_book_theme/assets, so I am a bit confused sphinx-theme-builder.readthedocs.io/en/latest/reference/filesystem-layout. Also, I think we should add the creation of *.mo files in webpack instead of committing them?

Hurrah! I wanted to make it easier / explicit for folks to keep compiled assets and uncompiled assets separate -- and I am happy to see that this caught non-CSS/JS stuff here! :)

> where do you suggest something like a master/sphinx_book_theme/translations folder 

I'd suggest leaving them under `theme/sphinx_book_theme/` for now, and then coming back around to cleaning that up in a follow up PR.

---

My heuristic for thinking about such things is: Would it be less work for reviewers if this was in a separate PR? In this case, I think the answer is a yes. 

It's _technically_ where they are on `main` at the moment as well (so... in theory, someone could do `{% include "sphinx_book_theme/translations/.gitignore" %}` in their templates[^1]). We can certainly clean this up in well-scoped follow up PR. It makes reviewing easier, and, gives you a clearer list of "what all did I end up doing". :)

[^1]: Yes, that's why we should move those files out! :)

**Comment by choldgraf** [comment]
I think @pradyunsg has a good intuition here - we should get this PR into a mergeable state ASAP, and we can always use follow-up PRs to clean things up if need be - those won't be nearly as disruptive from a merge conflict standpoint.

**Comment by AakashGfude** [comment]
Yes, totally agree with doing it in follow-up PRs. I guess mostly wanted to bring this issue to light here. Will create a separate issue once this is merged.  

I have moved the `translations` folder to `src/sphinx_book_theme` instead of leaving them under `src/sphinx_book_theme/theme/sphinx_book_theme`. Because to make the translations module importable I had to create empty `__init__.py` in both `theme` folder and in `theme/sphinx_book_theme` folder (make them modules as well). To Make the dotted import possible. 

As we are importing `_convert` python file of translations for testing:
`from sphinx_book_theme.translations._convert import convert_json`

Let me know if I should revert it and this isn't following proper convention.

**Comment by AakashGfude** [comment]
- [ ]  Removed `theme_dev_mode`, config variable, as I think it's not needed anymore, given `sphinx-theme-builder`. https://github.com/executablebooks/sphinx-book-theme/pull/469/files#diff-827e5b7309121354fc6d5411c429a4a07800bc6abee0db2f47b60c400ec4be99L26

**Comment by AakashGfude** [comment]
Have no idea, how to fix the pre-commit.ci failure😢

**Comment by choldgraf** [comment]
Hmm maybe this is related to the "additional dependencies" config? Could we specify the extra dependency in the error message there?

https://github.com/mgedmin/check-manifest#version-control-integration

I wonder if @pradyunsg could weigh in on whether this is somehow related to the sphinx theme builder configuration setup 🎉

**Comment by pradyunsg** [comment]
You don't need check-manifest anymore. That said, that looks like something that's a bit of a Ubuntu mess related to breaking `ensurepip`?

**Comment by choldgraf** [comment]
Is that because we're not checking in programmatically generated assets into the repo anymore? If that's the case then I am +1 on removing it.

**Comment by chrisjsewell** [comment]
> Is that because we're not checking in programmatically generated assets into the repo anymore?

and by extension, what does this mean for `MANIFEST.in`?

File: pyproject.toml
Comment (chrisjsewell) [inline pyproject.toml:21]: Does this PR now mean docutils 0.17 is supported?

**Comment by pradyunsg** [comment]
That can be removed as well! sphinx-theme-builder determines what files to include by asking git!

https://github.com/pradyunsg/sphinx-theme-builder/blob/10450cfeb65bc3610a6708ee731069c707978fef/src/sphinx_theme_builder/_internal/distributions.py#L33

**Comment by chrisjsewell** [comment]
I guess that's similar to flit 👍

**Comment by pradyunsg** [comment]
Basically, the sdist ends up being the same as a `git archive` or whatever is the equivalent of downloading from the GitHub repository *after* generating the assets.

**Comment by pradyunsg** [comment]
Yup, basically the same as flit. I'm pretty sure I copy-pasted some code for that too.

**Comment by choldgraf** [comment]
(side comment - I think all of these more recent developments in Python packaging are pretty cool, thanks @pradyunsg for all of the work you're doing in that space :-) )

File: pyproject.toml
Comment (choldgraf) [reply pyproject.toml:21]: I suspect we'll need to change some CSS still, so I'd assume that it won't be supported in this PR and we shouldn't bump that dependency just yet. I'm working on some HTML/CSS refactoring that I think will add support for docutils .17, but will do that as a follow-up to this PR so that we don't change too many things at the same time.

Comment (chrisjsewell) [reply pyproject.toml:21]: Yeh exactly, I was pretty sure that was the case; if the CSS still only covers `div+class`, and not also semantic HTML, then it does not support docutils 0.17

**Comment by AakashGfude** [comment]
> Basically, the sdist ends up being the same as a `git archive` or whatever is the equivalent of downloading from the GitHub repository _after_ generating the assets.

This is pretty cool, I will remove the MANIFEST file.

**Comment by AakashGfude** [comment]
The docs look good, apart from the margin directive: https://sphinx-book-theme--469.org.readthedocs.build/en/469/reference/special-theme-elements.html .
Because the margin element is `aside` instead of `div`. Will add styling for `aside`.

**Comment by choldgraf** [comment]
@AakashGfude the aside thing is broken because this PR increases the docutils version, I think

**Comment by AakashGfude** [comment]
@choldgraf have done a quick fix here, but will open an issue, to check style updates needed for docutils update.

- [ ] open issue to address style changes for docutils update.

**Comment by choldgraf** [comment]
@AakashGfude I think there's already an issue for it, so double check that one first. Though hopefully there's not a ton to change because we've been removing the "div" assumption here and there and just using class names

**Comment by AakashGfude** [comment]
Looks okayish to review. 
I still need to update the docs though.

File: src/sphinx_book_theme/__init__.py
Comment (choldgraf) [inline src/sphinx_book_theme/__init__.py:14]: should we just delete this?

File: src/sphinx_book_theme/__init__.py
Comment (choldgraf) [inline src/sphinx_book_theme/__init__.py:234]: Same here - can we just remove this?

File: src/sphinx_book_theme/assets/styles/_footer.scss
Comment (choldgraf) [inline src/sphinx_book_theme/assets/styles/_footer.scss:5]: Just to confirm - have you changed anything in the CSS files other than either moving them or updating their formatting in a way that doesn't change end behavior? If you *have* changed something, can you briefly describe what you've changed?

File: tox.ini
Comment (choldgraf) [inline tox.ini:14]:
```suggestion
envlist = py39-sphinx3
```

Perhaps we should stick with the latest supported Python version as the default tox environment?

File: webpack.config.js
Comment (choldgraf) [inline webpack.config.js:70]: Could you add a one-line comment for each of these explaining what it does (very briefly)?

**Review by choldgraf — COMMENTED** [review]
This looks really nice to me! I checked out this PR locally and was able to successfully run `tox -e docs-live` 🎉 

I've left a few (relatively minor) comments throughout, let me know what you think.

I agree that the main thing left to do is document this infrastructure so that others know what's going on. I don't think we want to block this PR on having a ton of docs. How about:

- Provide some basic information about how to build the theme locally with tox
- Provide some basic information about the sphinx-theme-builder
- Provide some basic information about the webpack and JS-specific files that you've added so others know what's going on

You can use the [pydata contributor documentation](https://pydata-sphinx-theme.readthedocs.io/en/latest/contribute/setup.html) for inspiration about what stuff to include.

Once we have those basic pieces there, so that the contributing documentation is "technically correct" even if not totally complete, then I think we should merge this in and start iterating in more focused PRs from there.

File: src/sphinx_book_theme/assets/styles/_footer.scss
Comment (AakashGfude) [reply src/sphinx_book_theme/assets/styles/_footer.scss:5]: I have added a few styles, for the margin directive in this commit https://github.com/executablebooks/sphinx-book-theme/pull/469/commits/4fb51549f59c6075ba5e1eb185ded991a244031e , which I pointed out in this comment https://github.com/executablebooks/sphinx-book-theme/pull/469#issuecomment-1019616614. Still need to find the issue which was created for this. I will update that issue once I find it.

File: tox.ini
Comment (AakashGfude) [reply tox.ini]: Sounds good

**Comment by AakashGfude** [comment]
> * Provide some basic information about how to build the theme locally with tox
> * Provide some basic information about the sphinx-theme-builder
> * Provide some basic information about the webpack and JS-specific files that you've added so others know what's going on

@choldgraf I updated the contributing guide https://github.com/executablebooks/sphinx-book-theme/pull/469/commits/fb5ed0adfcbd84cb5e658ccc25951334c2e1dfe1 , which briefly touches or just mentions all these points. I reckon we should add more, but we can do it in a follow-up PR?

I will check the translations README, and update it as well.

File: webpack.config.js
Comment (AakashGfude) [reply webpack.config.js:70]: Done

**Comment by choldgraf** [comment]
Those contributing docs look good to me 👍 - anything else you want to tackle in this PR? Otherwise I'd say we leave it open until tomorrow in case others want to make comments, and merge it in unless there are any objections. This PR touches a lot of stuff so it'd be good to get it in so that we can start iterating.

**Comment by AakashGfude** [comment]
- [ ] Add @choldgraf's email address

**Comment by AakashGfude** [comment]
Sounds good @choldgraf. I think I am done with any edits/additions in this PR.

File: src/sphinx_book_theme/assets/scripts/index.js
Comment (choldgraf) [inline src/sphinx_book_theme/assets/scripts/index.js:1]:
```suggestion
// Import CSS variables
// ref: https://css-tricks.com/getting-javascript-to-talk-to-css-and-sass/
import "../styles/index.scss";
```

**Comment by choldgraf** [comment]
OK, I took another pass and added in one quick comment, there are likely still some things that we need to clean up. The styles look good to me in general, but there seem to be a few tweaks to make (maybe related to docutils). I will tackle those in the refactor PR (#471) and we can hold off on closing #392 until that PR is merged as well.

I'll merge this PR so that we can start iterating a bit from here.

**Review by choldgraf — APPROVED** [review]

**Review by pradyunsg — COMMENTED** [review]
Hah. I had clicked the wrong button.

File: docs/contributing.md
Comment (choldgraf) [reply docs/contributing.md:111]: good point - I can take a pass at the contributing docs in general as well (or spot check other changes as needed) in https://github.com/executablebooks/sphinx-book-theme/pull/472

Comment (AakashGfude) [reply docs/contributing.md:111]: Thanks, @pradyunsg for pointing this out.

**Pull request merged**
